### PR TITLE
endpoint: avoid deferring code which locks endpoint in regenerateBPF

### PR DESCRIPTION
**Pull request by ianvernon**
If a panic occurs while this function is running and the endpoint lock is held
when the panic happens, the deferred proxy finalization / reverting code will
try to acquire the endpoint lock, resulting in the agent to hang. Now, the
function is returned to the caller of regenerateBPF to call when it returns.
This allows for a 'deferring' of the code in a sense, but the returned function
will only be called if a panic has not occurred.

More information about panic: https://blog.golang.org/defer-panic-and-recover

Signed-off by: Ian Vernon <ian@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6210)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Related to #6183, #6209

File: pkg/endpoint/bpf.go
Comment (tgraf) [inline pkg/endpoint/bpf.go]: The comment must say that the function must be invoked when it is non-nil.

**Review by tgraf — DISMISSED** [review]

**Review by aanm — CHANGES_REQUESTED** [review]
Unfortunately it won't be this simple, although the "`afterFunc`" is not executed in case of a panic, the caller of `regenerateBPF` has a `defer` function which also does a endpoint lock. So we will have the same problem.

Although I don't like it, I don't see a better way to do it than actually panic the whole program, if a panic happens we will panic hard and we need to fix why the panic happen which it is better than having the program in an unknown state. The way I suggest to fix this is to create a helper that can/ or probably should be used at the begging of all defer functions that does the following:

```golang
func PanicIfPanic() {
		if r := recover(); r != nil {
			panic(r)
		}
}
```
```golang
defer func() {
		PanicIfPanic()
		// lock endpoint ....
}
```

This way we won't deadlock because we don't stop the panic while doing a panic

**Comment by tgraf** [comment]
@aanm This is getting a bit insane, we can't expect a panic in every defer. I think #6214 is the right next step. I would like to see a helper for the metric manipulation though that checks for negative values and does a recover(). Metrics and logs are one of those things that we depend on from a lot of code paths and those things should simply never panic. It's insane.

**Comment by ianvernon** [comment]
So the upshot here is that this won't fix the issue. However, I still think that if we do get a panic anywhere in the code, and we have this `defer` logic which locks the endpoint, then we will hit the same problem. This indicates that there are flaws in how we are using `defer` and thus how we are writing some functions in our code. IMO we should audit the code and see what code is doing an endpoint lock in a defer statement and see how we can remove it. This is primarily due to the fact that we have several large functions which lock and unlock the endpoint multiple times, that also include deferred functions which lock the endpoint structure. We should split up these functions into smaller, more modular functions which defer the unlocking of the endpoint inside of them.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
I audited the code and only found one other deferred function which tries to acquire the endpoint lock, which occurs in `endpoint.regenerate`. I created a new type, `regenerationResult`, which holds a stack of functions to be ran after regeneration. The responsibility of invoking the functions in this stack should be done by the caller.

Let me know what you guys think about the design. I can add additional comments / cleanup / documentation once I get approval of the overall idea.

File: pkg/endpoint/bpf.go
Comment (ianvernon) [reply pkg/endpoint/bpf.go]: Done.

**Comment by ianvernon** [comment]
test-me-please

**Comment by aanm** [comment]
To be honest this seems to overkill and complex just to avoid deadlocking when a panic occurs. What's wrong the approach in https://github.com/cilium/cilium/pull/6210#pullrequestreview-173418163? It seems easier to backport as well.

**Comment by ianvernon** [comment]
@aanm I'm not a huge fan of this, but I think adding something like `PanicIfPanic` means that there is a certain way that code is being structured which doesn't mesh well with certain conditions (i.e., getting a `panic`). I think that we could backport the `PanicIfPanic` to releases that already have gone out to avoid a lot of churn (especially in a function as critical as `regenerateBPF`), but that in master, we should restructure the code to avoid the need for something like `PanicIfPanic` all together. I believe the fact that we hit this problem means that the code is not ideally structured. I've put this back in `wip` state in favor of #6191 for the time-being.

**Comment by ianvernon** [comment]
Closing in favor of #6191 for now.

**Comment by ianvernon** [comment]
Not deleting the branch just yet.

**Pull request closed without merging**
